image_client/picturae_importer.py

```diff
@@ -0,0 +1,986 @@
+"""this takes the csv data from picturae_csv_create, and parses the fields on a row by row basis into new db records,
+    filling in the necessary relational tables, and uploading attached images to image server.
+    Used as part of picturae_project for upload post-imaging, to create records to be updated post OCR and transcription
+"""
+import atexit
+import traceback
+import picturae_config
+from rpy2 import robjects
+from rpy2.robjects import pandas2ri
+from uuid import uuid4
+from picturae_import_utils import *
+import logging
+from string_utils import *
+from sql_csv_utils import *
+from importer import Importer
+from picturae_csv_create import CsvCreatePicturae
+from sql_csv_utils import generate_token
+from botany_importer import BotanyImporter
+
+from picturae_csv_create import starting_time_stamp
+
+
+class PicturaeImporter(Importer):
+    """DataOnboard:
+           A class with methods designed to wrangle, verify,
+           and upload a csv file containing transcribed
+           specimen sheet records into the casbotany database,
+           along with attached images
+    """
+
+    def __init__(self, paths, date_string=None):
+        super().__init__(picturae_config, "Botany")
+
+        self.date_use = date_string
+
+        self.logger = logging.getLogger('PicturaeImporter')
+        # full collector list is for populating existing and missing agents into collector table
+        # new_collector_list is only for adding new agents to agent table.
+        empty_lists = ['barcode_list', 'image_list', 'full_collector_list', 'new_collector_list',
+                       'taxon_list']
+
+        for empty_list in empty_lists:
+            setattr(self, empty_list, [])
+
+        self.no_match_dict = {}
+
+        # running csv create
+        CsvCreatePicturae(date_string=self.date_use)
+
+        self.file_path = f"PIC_upload/PIC_record_{self.date_use}.csv"
+
+        self.record_full = pd.read_csv(self.file_path)
+
+        self.batch_size = len(self.record_full)
+
+        # intializing parameters for database upload
+        init_list = ['GeographyID', 'taxon_id', 'barcode',
+                     'verbatim_date', 'start_date', 'end_date',
+                     'collector_number', 'locality', 'collecting_event_guid',
+                     'collecting_event_id', 'locality_guid', 'agent_guid',
+                     'geography_string', 'GeographyID', 'locality_id',
+                     'full_name', 'tax_name', 'locality',
+                     'determination_guid', 'collection_ob_id', 'collection_ob_guid',
+                     'name_id', 'author_sci', 'family', 'gen_spec_id', 'family_id', 'parent_author']
+
+        for param in init_list:
+            setattr(self, param, None)
+
```

Comment:
I probably hardcode this too, but it really ought to be in a config file. Generally speaking, if you've ever got a specific value, it belongs in a config file.

Reply:
I am not really sure how I would put this many parameters in a config file without getting confusing.